### DISCUSSION TITLE
scv.pl.scatter() error and gene phase portraits

### DISCUSSION THREAD

**Discussion by AAA-3** [theislab/scvelo#658] [Q&A]
Hello folks!

I am having issues with my data and code which I was hoping to get some assistance on - Issues listed are simillar to the points raised in #657  and the simillar error type as #199

I have the following code:
```
scv.pl.velocity(WT, ['Sox11',  'Tcf4', 'Nkx2-1', 'Dlx1'], ncols=2)
scv.pl.scatter(WT, 'Nkx2-1', color=['Clustername', 'velocity'],add_outline='Interneurons newborn, Interneurons 2,  Sst+ interneurons') 
```
<details>
<summary>Selected velocity plots are not giving the "football" curve that I have seen people talk about...can I use Velocity analysis with such a dataset?</summary>
<br>

![image](https://user-images.githubusercontent.com/68068335/130589640-b7e45350-743e-47cb-b2af-e5fb970a97ee.png)

</details>
 

<details>
<summary>Also, in following the scVelo documentation and trying to implement the second line of code, I get the following error...any clue why?</summary>
<br>

```
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_111153/516572165.py in <module>
      1 scv.pl.velocity(WT, ['Sox11',  'Tcf4', 'Nkx2-1', 'Dlx1'], ncols=2)
----> 2 scv.pl.scatter(WT, 'Nkx2-1', color=['Clustername', 'velocity'],add_outline='Interneurons newborn, Interneurons 2,  Sst+ interneurons')

~/anaconda3/lib/python3.8/site-packages/scvelo/plotting/scatter.py in scatter(adata, basis, x, y, vkey, color, use_raw, layer, color_map, colorbar, palette, size, alpha, linewidth, linecolor, perc, groups, sort_order, components, projection, legend_loc, legend_loc_lines, legend_fontsize, legend_fontweight, legend_fontoutline, xlabel, ylabel, title, fontsize, figsize, xlim, ylim, add_density, add_assignments, add_linfit, add_polyfit, add_rug, add_text, add_text_pos, add_outline, outline_width, outline_color, n_convolve, smooth, rescale_color, color_gradients, dpi, frameon, zorder, ncols, nrows, wspace, hspace, show, save, ax, **kwargs)
    182                     )
    183                 ax.append(
--> 184                     scatter(
    185                         adata,
    186                         ax=pl.subplot(gs),

~/anaconda3/lib/python3.8/site-packages/scvelo/plotting/scatter.py in scatter(adata, basis, x, y, vkey, color, use_raw, layer, color_map, colorbar, palette, size, alpha, linewidth, linecolor, perc, groups, sort_order, components, projection, legend_loc, legend_loc_lines, legend_fontsize, legend_fontweight, legend_fontoutline, xlabel, ylabel, title, fontsize, figsize, xlim, ylim, add_density, add_assignments, add_linfit, add_polyfit, add_rug, add_text, add_text_pos, add_outline, outline_width, outline_color, n_convolve, smooth, rescale_color, color_gradients, dpi, frameon, zorder, ncols, nrows, wspace, hspace, show, save, ax, **kwargs)
    615                         add_outline = f"{adata[:, basis].var[add_outline][0]}"
    616                 idx = groups_to_bool(adata, add_outline, color)
--> 617                 if idx is not None and np.sum(idx) > 0:  # if anything to be outlined
    618                     zorder = 2 if zorder is None else zorder + 2
    619                     if kwargs["s"] is not None:

<__array_function__ internals> in sum(*args, **kwargs)

~/anaconda3/lib/python3.8/site-packages/numpy/core/fromnumeric.py in sum(a, axis, dtype, out, keepdims, initial, where)
   2257         return res
   2258 
-> 2259     return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims,
   2260                           initial=initial, where=where)
   2261 

~/anaconda3/lib/python3.8/site-packages/numpy/core/fromnumeric.py in _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs)
     84                 return reduction(axis=axis, out=out, **passkwargs)
     85 
---> 86     return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
     87 
     88 

TypeError: cannot perform reduce with flexible type
```

![image](https://user-images.githubusercontent.com/68068335/130590164-031dc215-f162-48d9-8ba6-60ef7a77fe82.png)

</details>

**Comment by mxposed** [comment]
I'm not a maintainer, but hopefully can suggest something useful.

For the genes to look at with the footboll-shape curve, try looking at top-likelihood genes, as described in [tutorial](https://scvelo.readthedocs.io/DynamicalModeling/#Top-likelihood-genes). From what I understand, having such genes in a dataset is not guaranteed and maybe rare, so by looking at random genes you might not find that shape.

For the error: try removing `add_outline` parameter, and then adding back cell types one by one. Also, maybe the spaces around the commas are throwing it off.

**Reply by AAA-3** [reply]
Thanks for the advice! By "other genes" I did mean the genes ranked by velocity - some were more in line with requirements than others:

<details>
<summary>Ranked genes phase portraits</summary>
<br>
top two genes for each cluster (not in any order):

![image](https://user-images.githubusercontent.com/68068335/130791875-aa911462-b882-44a7-b50d-3b5ddacbf443.png)

</details>

Regarding the `scv.pl.scatter` error, I still am having difficulty. I think it may be to do with the way I might have formatted the column when I was importing seurat metadata into the anndata file earlier. Running the following code from the tutorial also gives me the same error:

<details>
<summary>Code and error:</summary>
<br>

```
kwargs = dict(frameon=False, size=10, linewidth=1.5,
              add_outline='Interneurons newborn,Interneurons 2,Sst+ interneurons')
scv.pl.scatter(WT, df['Interneurons newborn'][:5], ylabel='Interneurons newborn', **kwargs)
```

 Error:

```

TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_111153/4161561869.py in <module>
      3               add_outline='Interneurons newborn,Interneurons 2,Sst+ interneurons')
      4 
----> 5 scv.pl.scatter(WT, df['Interneurons newborn'][:5], ylabel='Interneurons newborn', **kwargs)
      6 scv.pl.scatter(WT, df['Interneurons 2'][:5], ylabel='Interneurons 2', **kwargs)
      7 scv.pl.scatter(WT, df['Sst+ interneurons'][:5], ylabel='Sst+ interneurons', **kwargs)

~/anaconda3/lib/python3.8/site-packages/scvelo/plotting/scatter.py in scatter(adata, basis, x, y, vkey, color, use_raw, layer, color_map, colorbar, palette, size, alpha, linewidth, linecolor, perc, groups, sort_order, components, projection, legend_loc, legend_loc_lines, legend_fontsize, legend_fontweight, legend_fontoutline, xlabel, ylabel, title, fontsize, figsize, xlim, ylim, add_density, add_assignments, add_linfit, add_polyfit, add_rug, add_text, add_text_pos, add_outline, outline_width, outline_color, n_convolve, smooth, rescale_color, color_gradients, dpi, frameon, zorder, ncols, nrows, wspace, hspace, show, save, ax, **kwargs)
    182                     )
    183                 ax.append(
--> 184                     scatter(
    185                         adata,
    186                         ax=pl.subplot(gs),

~/anaconda3/lib/python3.8/site-packages/scvelo/plotting/scatter.py in scatter(adata, basis, x, y, vkey, color, use_raw, layer, color_map, colorbar, palette, size, alpha, linewidth, linecolor, perc, groups, sort_order, components, projection, legend_loc, legend_loc_lines, legend_fontsize, legend_fontweight, legend_fontoutline, xlabel, ylabel, title, fontsize, figsize, xlim, ylim, add_density, add_assignments, add_linfit, add_polyfit, add_rug, add_text, add_text_pos, add_outline, outline_width, outline_color, n_convolve, smooth, rescale_color, color_gradients, dpi, frameon, zorder, ncols, nrows, wspace, hspace, show, save, ax, **kwargs)
    615                         add_outline = f"{adata[:, basis].var[add_outline][0]}"
    616                 idx = groups_to_bool(adata, add_outline, color)
--> 617                 if idx is not None and np.sum(idx) > 0:  # if anything to be outlined
    618                     zorder = 2 if zorder is None else zorder + 2
    619                     if kwargs["s"] is not None:

<__array_function__ internals> in sum(*args, **kwargs)

~/anaconda3/lib/python3.8/site-packages/numpy/core/fromnumeric.py in sum(a, axis, dtype, out, keepdims, initial, where)
   2257         return res
   2258 
-> 2259     return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims,
   2260                           initial=initial, where=where)
   2261 

~/anaconda3/lib/python3.8/site-packages/numpy/core/fromnumeric.py in _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs)
     84                 return reduction(axis=axis, out=out, **passkwargs)
     85 
---> 86     return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
     87 
     88 

TypeError: cannot perform reduce with flexible type
```

</details>

Google tells me the error has to do with python trying to read numbers that may have been stored as a `string` instead of `float` or `integer`... just not sure which set of numbers the error is referring to, since I assume the numbers it needs are ones computed and stored by scvelo.

**Reply by WeilerP** [reply]
@AAA-3, based on the phase portraits, you are showing, your dataset does not seem appropriate for velocity analysis.

Regarding the error message: AON, I can only tell you the same as your Google search suggests: I suspect the root of the issue is how  you preprocessing the data outside of scvelo (or Python) but cannot say for certain. In case you can share the data, would you please open an issue (providing all required info in the issue template s.t. I can simply copy and paste the code snippet) and provide the data? That way, I can track down the issue.

**Reply by AAA-3** [reply]
Hmmm thanks for that insight. I will have a think about  what you have said. 

I used stochastic mode on my data which is sourced from embryonic brain and so the cells may not have reached a steady state. I will check if I get anything different when I run it in dynamic mode. I will also look at the the genes listed in `fit_liklihood` instead of `rna_velocity` to see if that makes a difference. If at the end of it, if I still do not get the nice football curve I will shift to CellRank for fate decisions of the cells (#657). Before this, I will see if subsetting out the outlying clusters (#462) makes a difference. Does this seem like a suitable plan? There might have little to change but might as well give it a go.
